File: app/src/main/java/jp/juggler/subwaytooter/api/entity/TootRelationShip.java
package jp.juggler.subwaytooter.api.entity;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import jp.juggler.subwaytooter.table.UserRelation;
import jp.juggler.subwaytooter.util.LogCategory;
import jp.juggler.subwaytooter.util.Utils;

public class TootRelationShip {
	
	private static final LogCategory log = new LogCategory( "TootRelationShip" );
	
	// Target account id
	public long id;
	
	// Whether the authorized user is currently following the target account.
	// maybe faked in response of follow-request.
	public boolean following;
	
	// Whether the authorized user is currently being followed by the target account.
	public boolean followed_by;
	
	// Whether the authorized user is currently blocking the target account.
	public boolean blocking;
	
	// Whether the authorized user is currently muting the target account.
	public boolean muting;
	
	// Whether the authorized user has requested to follow the target account.
	// maybe true while follow-request is progress on server, even if the user is not locked.
	public boolean requested;
	
	// (mastodon 2.1 or later) per-following-user setting.
	// Whether the boosts from target account will be shown on authorized user's home TL.
	public int showing_reblogs = UserRelation.REBLOG_UNKNOWN;
	
	
	
	@Nullable
	public static TootRelationShip parse( JSONObject src ){
		if( src == null ) return null;
		try{
			TootRelationShip dst = new TootRelationShip();
			dst.id = Utils.optLongX( src, "id" );
			
			Object ov = src.opt( "following" );
			if( ov instanceof JSONObject ){
				// https://github.com/tootsuite/mastodon/issues/5856
				// 一部の開発版ではこうなっていた
				
				dst.following = true;
				
				ov = ( (JSONObject) ov ).opt( "reblogs" );
				if( ov instanceof Boolean ){
					dst.showing_reblogs = (Boolean) ov ? UserRelation.REBLOG_SHOW : UserRelation.REBLOG_HIDE;
				}else{
					dst.showing_reblogs = UserRelation.REBLOG_UNKNOWN;
				}
				
			}else{
				// 2.0 までの挙動
				dst.following = ( ov instanceof Boolean ? (Boolean) ov : false );
				
				// 2.1 の挙動
				ov = src.opt( "showing_reblogs" );
				if( dst.following && ov instanceof Boolean ){
					dst.showing_reblogs = (Boolean) ov ? UserRelation.REBLOG_SHOW : UserRelation.REBLOG_HIDE;
				}else{
					dst.showing_reblogs = UserRelation.REBLOG_UNKNOWN;
				}
			}
			
			dst.followed_by = src.optBoolean( "followed_by" );
			dst.blocking = src.optBoolean( "blocking" );
			dst.muting = src.optBoolean( "muting" );
			dst.requested = src.optBoolean( "requested" );
			return dst;
		}catch( Throwable ex ){
			log.trace( ex );
			log.e( ex, "TootRelationShip.parse failed." );
			return null;
		}
	}
	
	public static class List extends ArrayList< TootRelationShip > {
		public List(){
			super();
		}
		
		public List( int capacity ){
			super( capacity );
		}
	}
	
	@NonNull public static List parseList( JSONArray array ){
		List result = new List();
		if( array != null ){
			int array_size = array.length();
			result.ensureCapacity( array_size );
			for( int i = 0 ; i < array_size ; ++ i ){
				JSONObject src = array.optJSONObject( i );
				if( src == null ) continue;
				TootRelationShip item = parse( src );
				if( item != null ) result.add( item );
			}
		}
		return result;
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/api/entity/TootRelationShip.kt
package jp.juggler.subwaytooter.api.entity

import jp.juggler.subwaytooter.api.TootParser
import jp.juggler.subwaytooter.table.UserRelation
import jp.juggler.util.JsonObject

class TootRelationShip(parser:TootParser,src : JsonObject) {
	
	// Target account id
	// MisskeyのユーザリレーションはuserIdを含まないので後から何か設定する必要がある
	var id : EntityId
	
	// Whether the authorized user is currently following the target account.
	// maybe faked in response of follow-request.
	val following : Boolean
	
	// Whether the authorized user is currently being followed by the target account.
	val followed_by : Boolean
	
	// Whether the authorized user is currently blocking the target account.
	val blocking : Boolean

	// misskeyとMastodon 2.8.0以降
	val blocked_by : Boolean
	
	// Whether the authorized user is currently muting the target account.
	val muting : Boolean
	
	// Whether the authorized user has requested to follow the target account.
	// maybe true while follow-request is progress on server, even if the user is not locked.
	val requested : Boolean
	
	// (mastodon 2.1 or later) per-following-user setting.
	// Whether the boosts from target account will be shown on authorized user's home TL.
	val showing_reblogs : Int
	
	// 「プロフィールで紹介する」「プロフィールから外す」
	val endorsed : Boolean
	
	// misskey用
	val requested_by : Boolean
	
	// (Mastodon 3.2)
	var note : String? = null

	init {
		
		if( parser.serviceType == ServiceType.MISSKEY){
			this.id = EntityId.DEFAULT
			
			following = src.optBoolean("isFollowing")
			followed_by = src.optBoolean("isFollowed")
			muting = src.optBoolean("isMuted")
			blocking = src.optBoolean("isBlocking")
			blocked_by = src.optBoolean("isBlocked")
			requested = src.optBoolean("hasPendingFollowRequestFromYou")
			requested_by = src.optBoolean("hasPendingFollowRequestToYou")
			
			endorsed = false
			showing_reblogs = UserRelation.REBLOG_UNKNOWN
			
		}else{
			this.id = EntityId.mayDefault( src.string("id") )
			
			var ov = src["following"]
			if(ov is JsonObject) {
				// https://github.com/tootsuite/mastodon/issues/5856
				// 一部の開発版ではこうなっていた
				
				this.following = true
				
				ov = ov["reblogs"]
				if(ov is Boolean) {
					this.showing_reblogs = if(ov) UserRelation.REBLOG_SHOW else UserRelation.REBLOG_HIDE
				} else {
					this.showing_reblogs = UserRelation.REBLOG_UNKNOWN
				}
			} else {
				// 2.0 までの挙動
				this.following = if(ov is Boolean) ov else false
				
				// 2.1 の挙動
				ov = src["showing_reblogs"]
				if(this.following && ov is Boolean) {
					this.showing_reblogs = if(ov) UserRelation.REBLOG_SHOW else UserRelation.REBLOG_HIDE
				} else {
					this.showing_reblogs = UserRelation.REBLOG_UNKNOWN
				}
				
			}
			
			this.followed_by = src.optBoolean("followed_by")
			this.blocking = src.optBoolean("blocking")
			this.muting = src.optBoolean("muting")
			this.requested = src.optBoolean("requested")
			this.endorsed = src.optBoolean("endorsed")
			this.note = src.optString( "note")
			
			// https://github.com/tootsuite/mastodon/commit/9745de883b198375ba23f7fde879f6d75ce2df0f
			// Mastodon 2.8.0から
			this.blocked_by = src.optBoolean("blocked_by")
			
			requested_by = false
		}
		
	}
}
